File: platforms/android/sdk/src/main/kotlin/dev/bota/sdk/BotaDeviceClient.kt
package dev.bota.sdk

import dev.bota.sdk.internal.DeviceRuntime
import dev.bota.sdk.internal.runCleanupActions
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock

public class BotaDeviceClient internal constructor() {
    public val devices: DeviceManager = DeviceManager()
    public val controls: DeviceControlManager = DeviceControlManager()
    public val wifi: WiFiManager = WiFiManager()
    public val provisioning: ProvisioningManager = ProvisioningManager()
    public val factoryReset: FactoryResetManager = FactoryResetManager()
    public val recordings: RecordingManager = RecordingManager()
    public val ota: OTAManager = OTAManager()
    public val logs: DeviceLogManager = DeviceLogManager()

    private val lifecycle = Mutex()
    private var runtime: DeviceRuntime? = null

    public suspend fun configure(configuration: BotaConfiguration) {
        lifecycle.withLock {
            if (runtime != null) return
            val configured = configuration.runtimeFactory()
            runtime = configured
            devices.attach(configured)
            controls.attach(configured)
            wifi.attach(configured)
            provisioning.attach(configured)
            factoryReset.attach(configured)
            recordings.attach(configured)
            ota.attach(configured)
            logs.attach(configured)
        }
    }

    public suspend fun destroy() {
        lifecycle.withLock {
            val configured = runtime ?: return
            try {
                runCleanupActions(
                    { logs.detach() },
                    { ota.detach() },
                    { recordings.detach() },
                    { wifi.detach() },
                    { factoryReset.detach() },
                    { provisioning.detach() },
                    { controls.detach() },
                    { devices.detach() },
                    { configured.close() },
                )
            } finally {
                runtime = null
            }
        }
    }

    public companion object {
        @JvmField
        public val shared: BotaDeviceClient = BotaDeviceClient()
    }
}

File: platforms/apple/Sources/BotaAppleSDK/BotaDeviceClient.swift
public final class BotaDeviceClient: @unchecked Sendable {
    public static let shared = BotaDeviceClient()

    public let devices: DeviceManager
    public let wifi: WiFiManager
    public let provisioning: ProvisioningManager
    public let factoryReset: FactoryResetManager
    public let recordings: RecordingManager
    public let ota: OTAManager
    public let logs: DeviceLogManager
    private let lifecycle = BotaClientLifecycle()

    public init() {
        devices = DeviceManager()
        wifi = WiFiManager()
        provisioning = ProvisioningManager()
        factoryReset = FactoryResetManager()
        recordings = RecordingManager()
        ota = OTAManager()
        logs = DeviceLogManager()
    }

    public func configure(_ configuration: BotaConfiguration = .init()) async throws {
        try await lifecycle.configure(
            configuration,
            devices: devices,
            wifi: wifi,
            provisioning: provisioning,
            factoryReset: factoryReset,
            recordings: recordings,
            ota: ota,
            logs: logs
        )
    }

    public func destroy() async {
        await lifecycle.destroy(
            devices: devices,
            wifi: wifi,
            provisioning: provisioning,
            factoryReset: factoryReset,
            recordings: recordings,
            ota: ota,
            logs: logs
        )
    }
}

private actor BotaClientLifecycle {
    private var runtime: DeviceRuntime?

    func configure(
        _ configuration: BotaConfiguration,
        devices: DeviceManager,
        wifi: WiFiManager,
        provisioning: ProvisioningManager,
        factoryReset: FactoryResetManager,
        recordings: RecordingManager,
        ota: OTAManager,
        logs: DeviceLogManager
    ) async throws {
        guard runtime == nil else { return }
        let runtime = try await configuration.runtimeFactory()
        self.runtime = runtime
        await devices.attach(runtime)
        await wifi.attach(runtime)
        await provisioning.attach(runtime)
        await factoryReset.attach(runtime)
        await recordings.attach(runtime)
        await ota.attach(runtime)
        await logs.attach(runtime)
    }

    func destroy(
        devices: DeviceManager,
        wifi: WiFiManager,
        provisioning: ProvisioningManager,
        factoryReset: FactoryResetManager,
        recordings: RecordingManager,
        ota: OTAManager,
        logs: DeviceLogManager
    ) async {
        guard runtime != nil else { return }
        await logs.detach()
        await ota.detach()
        await recordings.detach()
        await wifi.detach()
        await provisioning.detach()
        await factoryReset.detach()
        await devices.detach()
        runtime = nil
    }
}
